File: NinePSharp/messages/Tsymlink.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Tsymlink : IMessage
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tsymlink;
    public ushort Tag { get; }
    public uint Fid { get; }
    public string Name { get; }
    public string Symtgt { get; }
    public uint Gid { get; }

    public Tsymlink(uint size, ushort tag, uint fid, string name, string symtgt, uint gid)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        Name = name;
        Symtgt = symtgt;
        Gid = gid;
    }

    public Tsymlink(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Name = data.ReadString(ref offset);
        Symtgt = data.ReadString(ref offset);
        Gid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Tsymlink);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        span.WriteString(Name, ref offset);
        span.WriteString(Symtgt, ref offset);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Gid);
    }
}

File: NinePSharp.Messages.FSharp/messages/Tsymlink.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tsymlink =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Name: string
      Symtgt: string
      Gid: uint32 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tsymlink
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tsymlink span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            writeString this.Name span &offset
            writeString this.Symtgt span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Gid)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let name = readString data &offset
        let symtgt = readString data &offset
        let gid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        { Size = size; Tag = tag; Fid = fid; Name = name; Symtgt = symtgt; Gid = gid }

